### ISSUE TITLE
Namespace collisions: consider R6 instead of S3

### ISSUE THREAD

**Issue by hongooi73** [thebioengineer/dockyard#6]
Just as a matter of design: convenience functions like `run`, `install`, `copy_from_url` etc are good to have, but also highly generic. Because of this, there's a good chance that they will collide with exported functions in other packages. In fact `update` already collides with the function from stats.

Instead of S3, perhaps consider using R6 to define your dockerfile class. Then you can expose these functions as class methods without having to worry about namespace collisions.

How it would look:

```r
shiny_dockerfile <- Dockerfile$
    new()$
    from("rocker/r-ver:devel")$
    update()$
    install("...")$
    ...
```

instead of magrittr pipes, you use _method chaining_, which is basically the same idea expressed in OOP syntax.

**Comment by thebioengineer** [comment]
I agree, I didnt like the namespace collisions but was in more of a trial-error mindset when first developing.

Interesting idea with the R6. I personally like R6 objects and the edit in place it enables. I've not had method chaining work for me in the past, but definitely will look into it more.

**Comment by hongooi73** [comment]
You have to return `self` from the methods for chaining to work:

```r
Dockerfile <- R6Class("Dockerfile",
public=list(
    install=function(...)
    {
        # generate text to install stuff
        return(self)
    }
))
```
